### PR TITLE
Feature/input validation network 457

### PR DESCRIPTION
**Pull request by pc532627**
### PR Type

- Refactoring (no functional changes)
- Tests

### Description

Input validation adjusted for networks.py

### How Has This Been Tested?

New test_networks.py file added and all pass on local machine.

### Does this PR introduce a breaking change?

N/A

### Checklist before requesting a review
- [x] I have made sure that my PR is not a duplicate.
- [x] My code follows the style guidelines of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have performed a self-review of my code.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.

### CODE REVIEW

File: tests/unit/test_networks.py [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+import coreax.networks
+import coreax.util
+
+
+class TestTrainState(unittest.TestCase):
+    """
+    Tests related to :func:`create_train_state`.
+
+    .. note::
+        Learning rate is not tested as it is part of weight updates, not initialisation
```
Comment (tp832944) [inline tests/unit/test_networks.py:36]: Insert a blank line after a Sphinx directive.

File: tests/unit/test_networks.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+        """
+        Test create_train_state with a negative hidden dimension size.
+        """
+        score_network = coreax.networks.ScoreNetwork(
+            -self.hidden_dimension, self.data_dimension
+        )
+
+        # Create a train state with this network - we expect optax to catch the invalid
+        # hidden dimension argument and raise an appropriate error
+        with self.assertRaises(TypeError) as error_raised:
```
Comment (tp832944) [inline tests/unit/test_networks.py:60]: Use `assertRaisesRegex` to combine this and line 69. Similarly for the other tests.

No need to include the full error message to test against, just enough to be sure we've hit the right one. No idea how stable the error wordings are between package versions. If you do want to match the full message, enclose the regex in `^...$` to match the start and end of the expressions.

File: tests/unit/test_networks.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+                module=score_network,
+                learning_rate=self.learning_rate,
+                data_dimension=self.data_dimension,
+                optimiser=self.optimiser,
+            )
+
+        self.assertEqual(
+            error_raised.exception.args[0],
+            f"Shapes must be 1D sequences of concrete values of integer type, "
+            f"got ({self.data_dimension}, {1.0*self.hidden_dimension}).",
```
Comment (tp832944) [inline tests/unit/test_networks.py:116]: Space out around the multiplier: `{1.0 * self`

File: tests/unit/test_networks.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+        with self.assertRaises(TypeError):
+            coreax.networks.create_train_state(
+                random_key=self.state_key,
+                module=score_network,
+                learning_rate=self.learning_rate,
+                data_dimension=-self.data_dimension,
+                optimiser=self.optimiser,
+            )
+
+        # We don't check the exact message as it uses a work (non-negative without the
```
Comment (tp832944) [inline tests/unit/test_networks.py:138]: word

**Review by tp832944 — CHANGES_REQUESTED** [review]
Validation is fine. I've requested a few tweaks to the tests. Note also that some validation tests for test_coresubset and test_reduction are failing on my own computer - are these being fixed elsewhere?

File: tests/unit/test_networks.py (hunk 1 above)
Comment (pc532627) [reply tests/unit/test_networks.py:36]: Added

File: tests/unit/test_networks.py (hunk 2 above)
Comment (pc532627) [reply tests/unit/test_networks.py:60]: Agreed and updated - I now just check that it discusses non-negative indices

Comment (pc532627) [reply tests/unit/test_networks.py:60]: Updated other tests to reflect this

File: tests/unit/test_networks.py (hunk 4 above)
Comment (pc532627) [reply tests/unit/test_networks.py:138]: Updated

**Comment by pc532627** [comment]
> Validation is fine. I've requested a few tweaks to the tests. Note also that some validation tests for test_coresubset and test_reduction are failing on my own computer - are these being fixed elsewhere?

Thanks - the 3 test fails are being fixed on the branch all of these are merged into. In short, these branches still use the requirement files for the unit test runs. However, on my machine, I've updated the environment and JAX has changed one of the error messages. I'm going to merge main into that collective branch which will resolve that.

**Comment by tp832944** [comment]
> > Validation is fine. I've requested a few tweaks to the tests. Note also that some validation tests for test_coresubset and test_reduction are failing on my own computer - are these being fixed elsewhere?
> 
> Thanks - the 3 test fails are being fixed on the branch all of these are merged into. In short, these branches still use the requirement files for the unit test runs. However, on my machine, I've updated the environment and JAX has changed one of the error messages. I'm going to merge main into that collective branch which will resolve that.

All the more reason to not be too precise with the checking of errors.

**Pull request merged**